From d94837b93b9966390052d6746c16d069b9037d34 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Mon, 14 Sep 2026 10:26:23 +0800
Subject: [PATCH] user32/tests: Move GetDIBits() tests with display DCs from
 gdi32.

CW-Bug-Id: #27702
(cherry picked from commit 6cd2c2c3df21f42630f7cfea248df556c8fc3b8c)
---
 dlls/gdi32/tests/bitmap.c   | 81 +----------------------------------
 dlls/user32/tests/monitor.c | 85 ++++++++++++++++++++++++++++++++++++-
 2 files changed, 84 insertions(+), 82 deletions(-)

diff --git a/dlls/gdi32/tests/bitmap.c b/dlls/gdi32/tests/bitmap.c
index 40e5f2d7d2a..773e0fad5ee 100644
--- a/dlls/gdi32/tests/bitmap.c
+++ b/dlls/gdi32/tests/bitmap.c
@@ -2105,10 +2105,9 @@ static void test_GetDIBits(void)
         0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
         0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
     };
-    unsigned int depths[] = {1, 2, 4, 8, 15, 16, 24, 32};
     HBITMAP hbmp;
     BITMAP bm;
-    HDC hdc, mem_dc;
+    HDC hdc;
     int i, bytes, lines;
     BYTE buf[1024];
     char bi_buf[sizeof(BITMAPINFOHEADER) + sizeof(RGBQUAD) * 256];
@@ -2419,84 +2418,6 @@ static void test_GetDIBits(void)
     ok(!memcmp(buf, dib_bits_24, sizeof(dib_bits_24)), "DIB bits don't match\n");
     DeleteObject(hbmp);
 
-    /* Test a SDL3 behavior where GetDIBits() gets called twice consecutively to get bmiColors.
-     * However, tests show that the second call fails when hbmp is a compatible bitmap and the
-     * bitmap depth is not 1-bit or 32-bit. Tests for GetDIBits() when hbmp is a DIB are already
-     * covered by other tests */
-    for (i = 0; i < ARRAY_SIZE(depths); i++)
-    {
-        winetest_push_context("depth %u", depths[i]);
-
-        hbmp = CreateBitmap(1, 1, 1, depths[i], NULL);
-
-        /* Simulate SDL3 behavior by calling GetDIBits() twice */
-        memset(bi, 0, sizeof(*bi));
-        bi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-        lines = GetDIBits(hdc, hbmp, 0, 0, NULL, bi, DIB_RGB_COLORS);
-        ok(lines == 1, "GetDIBits failed.\n");
-        lines = GetDIBits(hdc, hbmp, 0, 0, NULL, bi, DIB_RGB_COLORS);
-        if (depths[i] == 1)
-        {
-            ok(lines == 1, "GetDIBits failed.\n");
-            ok(*(unsigned int *)bi->bmiColors == 0, "Got unexpected bmiColors %#x\n", *(unsigned int *)bi->bmiColors);
-        }
-        else if (depths[i] == 32)
-        {
-            ok(lines == 1, "GetDIBits failed.\n");
-            ok(*(unsigned int *)bi->bmiColors == 0xff0000, "Got unexpected bmiColors %#x\n", *(unsigned int *)bi->bmiColors);
-        }
-        else
-        {
-            ok(lines == 0, "GetDIBits succeeded.\n");
-            ok(*(unsigned int *)bi->bmiColors == 0, "Got unexpected bmiColors %#x\n", *(unsigned int *)bi->bmiColors);
-
-            /* lines > 0. Still fails */
-            lines = GetDIBits(hdc, hbmp, 0, 1, NULL, bi, DIB_RGB_COLORS);
-            ok(lines == 0, "GetDIBits failed.\n");
-
-            /* buf != NULL. Still fails */
-            lines = GetDIBits(hdc, hbmp, 0, 1, buf, bi, DIB_RGB_COLORS);
-            ok(lines == 0, "GetDIBits failed.\n");
-
-            /* Reset biBitCount to 0. Now it succeeds */
-            bi->bmiHeader.biBitCount = 0;
-            lines = GetDIBits(hdc, hbmp, 0, 0, NULL, bi, DIB_RGB_COLORS);
-            ok(lines == 1, "GetDIBits failed.\n");
-        }
-
-        /* Test that GetDIBits() is rejecting DDB that's not 1-bit or 32-bit */
-        memset(bi, 0, sizeof(*bi));
-        bi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-        bi->bmiHeader.biWidth = 1;
-        bi->bmiHeader.biHeight = 1;
-        bi->bmiHeader.biPlanes = 1;
-        bi->bmiHeader.biBitCount = depths[i];
-        lines = GetDIBits(hdc, hbmp, 0, 1, buf, bi, DIB_RGB_COLORS);
-        if (depths[i] == 1 || depths[i] == 32)
-            ok(lines == 1, "GetDIBits failed.\n");
-        else
-            ok(lines == 0, "GetDIBits succeeded.\n");
-
-        /* Same result when using a memory DC so it's not related the display DC */
-        mem_dc = CreateCompatibleDC(hdc);
-        SelectObject(mem_dc, hbmp);
-        memset(bi, 0, sizeof(*bi));
-        bi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-        bi->bmiHeader.biWidth = 1;
-        bi->bmiHeader.biHeight = 1;
-        bi->bmiHeader.biPlanes = 1;
-        bi->bmiHeader.biBitCount = depths[i];
-        lines = GetDIBits(mem_dc, hbmp, 0, 1, buf, bi, DIB_RGB_COLORS);
-        if (depths[i] == 1 || depths[i] == 32)
-            ok(lines == 1, "GetDIBits failed.\n");
-        else
-            ok(lines == 0, "GetDIBits succeeded.\n");
-
-        DeleteDC(mem_dc);
-        DeleteObject(hbmp);
-        winetest_pop_context();
-    }
-
     ReleaseDC(0, hdc);
 }
 
diff --git a/dlls/user32/tests/monitor.c b/dlls/user32/tests/monitor.c
index 41b9cd96ea2..b139d84debc 100644
--- a/dlls/user32/tests/monitor.c
+++ b/dlls/user32/tests/monitor.c
@@ -2650,11 +2650,11 @@ static void _check_display_dc(INT line, HDC hdc, const DEVMODEA *dm, BOOL allow_
 
 static void test_display_dc(void)
 {
-    static const INT bpps[] = {1, 4, 8, 16, 24, 32};
+    static const INT bpps[] = {1, 4, 8, 16, 24, 32}, ddb_bpps[] = {1, 2, 4, 8, 15, 16, 24, 32};
     unsigned char buffer[sizeof(BITMAPINFOHEADER) + 256 * sizeof(RGBQUAD)], bits_buffer[1024];
     HBITMAP hbitmap, hbitmap2, old_hbitmap;
     BITMAPINFO *bmi = (BITMAPINFO *)buffer;
-    INT count, old_count, i, bpp, value;
+    INT count, old_count, i, j, bpp, value;
     DWORD device_idx, mode_idx;
     DEVMODEA dm, dm2, dm3;
     HDC hdc, hdc2, mem_dc;
@@ -2855,6 +2855,87 @@ static void test_display_dc(void)
             DeleteObject(hbitmap);
 
             DeleteDC(mem_dc);
+
+            /* Test GetDIBits() with the display DC and DDBs of various bit depths */
+            for (j = 0; j < ARRAY_SIZE(ddb_bpps); j++)
+            {
+                winetest_push_context("DDB bpp %d", ddb_bpps[j]);
+                hbitmap = CreateBitmap(1, 1, 1, ddb_bpps[j], NULL);
+
+                /* Simulate SDL3 behavior by calling GetDIBits() twice */
+                memset(buffer, 0, sizeof(buffer));
+                bmi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+                count = GetDIBits(hdc2, hbitmap, 0, 0, NULL, bmi, DIB_RGB_COLORS);
+                ok(count == 1, "GetDIBits failed.\n");
+                count = GetDIBits(hdc2, hbitmap, 0, 0, NULL, bmi, DIB_RGB_COLORS);
+                if (ddb_bpps[j] == 1 || (bpps[i] == 8 && ddb_bpps[j] == 8))
+                {
+                    todo_wine_if(bpps[i] == 8 && ddb_bpps[j] == 8)
+                    ok(count == 1, "GetDIBits failed.\n");
+                    ok(*(unsigned int *)bmi->bmiColors == 0, "Got unexpected bmiColors %#x\n",
+                       *(unsigned int *)bmi->bmiColors);
+                }
+                else if (ddb_bpps[j] == 32)
+                {
+                    ok(count == 1, "GetDIBits failed.\n");
+                    ok(*(unsigned int *)bmi->bmiColors == 0xff0000, "Got unexpected bmiColors %#x\n",
+                       *(unsigned int *)bmi->bmiColors);
+                }
+                else
+                {
+                    ok(count == 0, "GetDIBits succeeded.\n");
+                    ok(*(unsigned int *)bmi->bmiColors == 0, "Got unexpected bmiColors %#x\n",
+                       *(unsigned int *)bmi->bmiColors);
+
+                    /* lines > 0. Still fails */
+                    count = GetDIBits(hdc2, hbitmap, 0, 1, NULL, bmi, DIB_RGB_COLORS);
+                    ok(count == 0, "GetDIBits succeeded.\n");
+
+                    /* buf != NULL. Still fails */
+                    count = GetDIBits(hdc2, hbitmap, 0, 1, bits_buffer, bmi, DIB_RGB_COLORS);
+                    ok(count == 0, "GetDIBits succeeded.\n");
+
+                    /* Reset biBitCount to 0. Now it succeeds */
+                    bmi->bmiHeader.biBitCount = 0;
+                    count = GetDIBits(hdc2, hbitmap, 0, 0, NULL, bmi, DIB_RGB_COLORS);
+                    ok(count == 1, "GetDIBits failed.\n");
+                }
+
+                /* Test that GetDIBits() is rejecting some DDB bit depths */
+                memset(buffer, 0, sizeof(buffer));
+                bmi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+                bmi->bmiHeader.biWidth = 1;
+                bmi->bmiHeader.biHeight = 1;
+                bmi->bmiHeader.biPlanes = 1;
+                bmi->bmiHeader.biBitCount = ddb_bpps[j];
+                count = GetDIBits(hdc2, hbitmap, 0, 1, bits_buffer, bmi, DIB_RGB_COLORS);
+                if (ddb_bpps[j] == 1 || (bpps[i] == 8 && ddb_bpps[j] == 8) || ddb_bpps[j] == 32)
+                    todo_wine_if(bpps[i] == 8 && ddb_bpps[j] == 8)
+                    ok(count == 1, "GetDIBits failed.\n");
+                else
+                    ok(count == 0, "GetDIBits succeeded.\n");
+
+                /* Same result when using a display compatible memory DC */
+                mem_dc = CreateCompatibleDC(hdc2);
+                SelectObject(mem_dc, hbitmap);
+                memset(buffer, 0, sizeof(buffer));
+                bmi->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+                bmi->bmiHeader.biWidth = 1;
+                bmi->bmiHeader.biHeight = 1;
+                bmi->bmiHeader.biPlanes = 1;
+                bmi->bmiHeader.biBitCount = ddb_bpps[j];
+                count = GetDIBits(mem_dc, hbitmap, 0, 1, bits_buffer, bmi, DIB_RGB_COLORS);
+                if (ddb_bpps[j] == 1 || (bpps[i] == 8 && ddb_bpps[j] == 8) || ddb_bpps[j] == 32)
+                    todo_wine_if(bpps[i] == 8 && ddb_bpps[j] == 8)
+                    ok(count == 1, "GetDIBits failed.\n");
+                else
+                    ok(count == 0, "GetDIBits succeeded.\n");
+                DeleteDC(mem_dc);
+
+                DeleteObject(hbitmap);
+                winetest_pop_context();
+            }
+
             ReleaseDC(0, hdc2);
         }
         winetest_pop_context();
